rust/vapid/src/lib.rs

```diff
@@ -0,0 +1,482 @@
+//! VAPID auth support
+//!
+//! This library only supports the latest VAPID-draft-02+ specification.
+//!
+//! Example Use:
+//! ```
+//!
+//! // Create a key from an existing EC Private Key PEM file.
+//! // You can generate this with
+//! // Key::generate().to_pem("pem/file/path.pem")?;
+//! let my_key = Key::from_pem("pem/file/path.pem")?;
+//!
+//! // Construct the Claims hashmap
+//! let mut claims:HashMap<String, serde_json::Value> = HashMap::new();
+//! claims.insert(
+//!     String::from("sub"), serde_json::Value::from("mailto:bob@example.com")
+//! );
+//! // while `exp` can be filled in for you, `aud` should point to the net location of the
+//! // Push server you wish to talk to. (e.g. `https://push.services.mozilla.org`)
+//! // `aud` is optional for Mozilla, but may be required for GCM/FCM or other systems.
+//! claims.insert(
+//!     String::from("aud"), serde_json::Value::from("https://host.ext")
+//! );
+//!
+//! // The result will contain the `Authorization:` header. How you inject this into your
+//! // request is left as an exercise.
+//! let authorization_header = sign(key, &mut claims)?;
+//!
+//! ```
+
+extern crate asn1;
+extern crate base64;
+extern crate openssl;
+extern crate serde_json;
+extern crate time;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::Read;
+use std::io::Write;
+use std::path::Path;
+
+use openssl::bn::BigNum;
+use openssl::bn::BigNumContext;
+use openssl::hash::MessageDigest;
+use openssl::ec;
+use openssl::nid;
+use openssl::sign::{Signer, Verifier};
+use openssl::pkey;
+
+pub struct Key(ec::EcKey);
+
+/// a Key is a helper for creating or using a VAPID EC key.
+///
+/// Vapid Keys are always Prime256v1 EC keys.
+///
+#[allow(dead_code)]
+impl Key {
+    fn new() -> Key {
+        Key(ec::EcKey::from_curve_name(Key::name()).expect(
+            "EC Prime256v1 curve not supported",
+        ))
+    }
+
+    fn name() -> nid::Nid {
+        nid::X9_62_PRIME256V1
+    }
+
+    fn group() -> ec::EcGroup {
+        ec::EcGroup::from_curve_name(Key::name()).expect("EC Prime256v1 curve not supported")
+    }
+
+    /// Read a VAPID private key stored in `path`
+    pub fn from_pem(path: String) -> Result<Key, String> {
+        let mut pem_file =
+            File::open(&Path::new(&path)).expect(&format!("Could not open file at {}", path));
+        let mut pem_data = String::new();
+        pem_file.read_to_string(&mut pem_data).expect(&format!(
```

Comment:
You'll probably want to use `map_err` here or some other error-handling mechanism other than a panic (`expect` panics by default) as in general in Rust a panic is a bug (sort of like an assert in C) rather than a normal runtime error (like an exception in Python I believe)

Reply:
There's a number of libraries available for managing errors today, the most prominent of which is likely [`error-chain`](https://crates.io/crates/error-chain). Although just returning a `String` is fine for a first past of course!

Reply:
As food for though though:

with error-chain this would look like:

```rust
pem_file.read_to_string(&mut pem_data).chain_err(|| {
    format!("failed to read file: {}", path)
})?; // <- note the `?` operator
```

The benefit of a "concrete error type" over a string error type mostly comes up in interoperation with other libraries, so when this crate is consumed by an application, for example, it'd be likely a little nicer to work with a `Result<Key, MyError>` rather than a `Result<Key, String>`. 

Again though that's very far-reaching stuff, I'd stick to strings for now as it's not too hard to refactor later!

Reply:
Ah, yes. I saw that in the autopush refactor but wasn't quite sure how it worked. I may look into converting, since it does seem a bit more elegant and preserves the original errors.